### PR TITLE
chore(cli): accept hooks

### PR DESCRIPTION
**Pull request by pi0**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

CLI hooks are primary usable for custom CLIs built on top of @nuxt/cli (nuxt-ts). We can pass a `hooks` object as an argument to `run` command to register hooks. Only two useful hooks added for now:

- `setup({ cmd, dev, argv })`: will be called after parsing arguments. Adding possibility to modify args, command, etc and also setup env (like registering ts-node)
- `config(options)`: will be called after the config is loaded. before passing to nuxt. Adding the possibility to modify config. Like adding some extensions or modules.

## Related issues

- nuxt/typescript#79
- nuxt/typescript#80

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

### CODE REVIEW

File: packages/cli/src/run.js [hunk 1]
```diff
@@ -33,12 +33,20 @@ export default async function run (_argv) {
     cmd = await getCommand('dev')
   }
 
+  // Check for dev
+  const dev = argv[0] === 'dev'
+
+  // Call setup hook
+  if (typeof hooks.setup === 'function') {
+    await hooks.setup({ cmd, dev, argv })
```
Comment (clarkdo) [inline packages/cli/src/run.js]: If setup hook has been called before hook registration, does it need to be registered again?

**Review by clarkdo — COMMENTED** [review]
Will cli hooks be able to be configured from nuxt.config?

File: packages/cli/src/run.js (hunk 1 above)
Comment (pi0) [reply packages/cli/src/run.js]: Indeed this is a special hook which will be called before command initialization and will be never called again We can remove it from object taught if you want :)

**Comment by pi0** [comment]
> Will cli hooks be able to be configured from nuxt.config?

@clarkdo  no. they are CLI level hooks that will be hooked before nuxt is being ready. As mentioned in PR description that's a internal use-case for better typescript support. (or advanced users that want to build a cli on top of nuxt)

File: packages/cli/src/run.js (hunk 1 above)
Comment (pi0) [reply packages/cli/src/run.js]: done

**Review by kevinmarrec — COMMENTED** [review]
@pi0 Nice ! But how do we easy get rootDir with `setup` Hook ?

**Comment by pi0** [comment]
@kevinmarrec Default value of `rootDir` is `path.resolve(argv._[0] || '.')` before loading nuxt.config.js. We can also access resolved value with `config` hook.

**Comment by kevinmarrec** [comment]
@pi0 but CLI options aren't already parsed so `argv._[0]` doesn't exist yet or am I missing something ?

**Comment by kevinmarrec** [comment]
Nvm I read again NuxtCommand and it should be alright, argv passed to hook is parsed right ?

**Comment by pimlie** [comment]
Not sure what the issue is with typescript, but wasnt the idea of NuxtCommand's that they are easily extendible by themselves? Eg hooks might just be a second way of changing the task of a command?

How much granularity do we want with hooks? Eg in Nuxt we often use `:before` / `:after` suffixes, should we do that here too to keep consistency (even when we dont want to add one of the suffixes)?

**Comment by pi0** [comment]
Not much granularity is required and nuxt-ts is not adding new commands but need a hook to:
- Access parsed args (and rootDir) before command executation to be able to register ts-node
- Extend options and add extensions for runtime

Both are internal use cases and may not be even documented.

**Comment by kevinmarrec** [comment]
Still if we refactor `nuxt-ts` it will only work with next Nuxt version, we'll need to bump `nuxt-ts` to a major version and document it right ?

**Comment by pi0** [comment]
@kevinmarrec the hooks param is an optional addition. It is like we have to say bump to 2.9.2 or 2.10.0 to fix this issue.

**Comment by kevinmarrec** [comment]
@pi0 Yeah I was more likely thinking about the `register` thing that could be really improved with the `setup` hooks. Like only using hooks :)

I mean we can't use only hooks if we want `@nuxtjs/runtime` still work if user doesn't have yet the Nuxt version that use hooks.

**Comment by pi0** [comment]
Yep. Now i see your point why bumping ts module a major version for hooks 😊

**Review by kevinmarrec — APPROVED** [review]

**Pull request merged**